dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -576,18 +571,21 @@ private String getFromSubQuery( TrackedEntityInstanceQueryParams params, boolean
      */
     private String getFromSubQuerySelect( TrackedEntityInstanceQueryParams params )
     {
-        LinkedHashSet<String> columns = new LinkedHashSet<>(
-            List.of( "TEI.trackedentityinstanceid", "TEI.uid", "TEI.created", "TEI.lastupdated",
-                "TEI.inactive", "TEI.trackedentitytypeid", "TEI.potentialduplicate", "TEI.deleted", "OU.uid as ou",
-                "OU.name as ouname " ) );
+        List<String> defaultSelectFields = List.of( "TEI.trackedentityinstanceid", "TEI.uid", "TEI.created",
+            "TEI.lastUpdated",
+            "TEI.inactive", "TEI.trackedentitytypeid", "TEI.potentialduplicate", "TEI.deleted", "OU.uid as ou",
+            "OU.name as ouname " );
+
+        LinkedHashSet<String> columns = new LinkedHashSet<>( defaultSelectFields );
 
         for ( OrderParam orderParam : params.getOrders() )
         {
             Optional<TrackedEntityInstanceQueryParams.OrderColumn> orderColumn = findColumn( orderParam.getField() );
 
-            if ( orderColumn.isPresent() )
+            if ( orderColumn.isPresent()
+                && !defaultSelectFields.contains( orderColumn.get().getSqlColumnWithTableAlias() ) )
```

Comment:
I don't think its a good idea to do these substring searches in loops. Where you not using LinkedHashSet explicitly to account for duplicates? 🤔 This should turn the operation into O(1). Same above in `getQuerySelect`.